keps/0033-structured-parameters.md

```diff
@@ -0,0 +1,432 @@
+
+A structured data structure provides a lot more possibilities to organize parameters for an operator. Current operators organize their parameters with prefixes, but this has its limits.
+
+Additionally, we want to enhance parameters with more metadata: 
+- Label
+- Description
+- Grouping of parameters
+- Validation
+- More parameter types (enums, numbers, etc.)
+- UI flags (important or low-prio and hidden by default)
+
+Some of these are already possible with the current `params.yaml` format, some need to be added.
+
+### Goals
+
+TODO
+
+### Non-Goals
+
+TODO
+
+## Proposal
+
+Adopt json-schema as the default format for `params.yaml`.
+
+- Json-Schema allows all of our requirements to be fulfilled
+- It is an existing standard
+- It can be expressed as yaml to fit into the Kubernetes ecosystem
+- It can be extended with custom keywords to define our custom attributes like `trigger` and `immutable`
+- There are existing go libraries 
+
+### Overview
+
+ - KUDO operator package format
+   The package format has to be adjusted. We need a new version, and KUDO should be able to read both old and new formats.
+- CRDS
+- Instance Updates `kudo update ... -p KEY=value`
+- Values.yaml `kudo update ... -P values.yaml` 
+- Add Parameter Wizard `kudo package add parameter`
+- Parameter Listing`kudo package list parameters`
+  We need a new output format here - maybe a tree-table?
+- Additional attributes in json-schema
+  We want to use json-schema to describe the parameter structure, but we need additional information that is not described json-schema.
+   
+### Package Format
+
+TODO: Describe new package format (json-schema)
+
+### CRDs
+
+Normally, per Kubernetes conventions, CRDs would evolve over released versions without a breaking change from one version to the next, allowing clients to gradually convert to new CRD versions.
+
+With this change, this will not be possible, as json-schema supports a lot of configuration which will not be mappable from a flat list.
+
+There will be a hard break from the current from `v1beta1` to the new `v1beta2`. We will provide a migration path for existing installed CRDs, but the clients accessing these CRDs will need to be updated.
+
+#### Conversion
+
+To migrate the existing CRDs, we will implement a Conversion Webhook that only allows `v1beta1` to `v1beta2` conversion, not the other way around. When KUDO is upgraded, it will:
+
+- Run Pre-Migration
+    - Fetch all existing CRDs, run the conversion and check for any errors
+- Deploy the new CRD version
+    - `v1beta1` will be still in the list, but with `storage=false, served=false`
+    - `v1beta2` will be marked as `storage=true, served=true`
+- Deploy the new manager with the conversion webhook
+- Run the migration
+    - Fetch all CRDs and issue an update - This will trigger a conversion and save the CRD with the new version
+    - Update the CRD status.storedVersions field
+- With the following KUDO release, we can remove `v1beta1` from the CRD list.
```

Comment:
It's great that we will provide a migration path but that's a very short migration period. What is stopping us from providing the conversion webhook for a few versions of KUDO to provide a migration period of, e.g., 6 months?

Reply:
Yeah, that's reasonable - There's no need to really speed up the removal. And you're right, as soon as we remove the old version the upgrade path is not available anymore.